text/0000-structural-records.md

```diff
@@ -0,0 +1,1208 @@
+- Feature Name: `stuctural_records`
+- Start Date: 2018-11-02
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+**Introduce _structural records_** of the form `{ foo: 1u8, bar: true }` of
+type `{ foo: u8, bar: bool }` into the language. Another way to understand
+these sorts of objects is to think of them as *"tuples with named fields"*,
+*"unnamed structs"*, or *"anonymous structs"*.
+
+# Motivation
+[motivation]: #motivation
+
+There are four motivations to introduce structural records,
+two major and two minor.
+
+## Major: Improving ergonomics, readability, and maintainability
+
+Sometimes, you just need a one-off struct that you will use a few times for
+some public API or in particular for some internal API. For example, you'd like
+to send over a few values to a function, but you don't want to have too many
+function arguments. In such a scenario, defining a new nominal `struct` which
+contains all the fields you want to pass is not a particularly ergonomic solution.
+Instead, it is rather heavy weight:
+
+```rust
+struct Color {
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+fn caller() {
+    ...
+    let color = Color { red: 255, green: 0, blue: 0 };
+    ...
+    do_stuff_with(color);
+    ...
+}
+
+fn do_stuff_with(color: Color) { // Only time we use `Color`!
+    some_stuff(color.red);
+    ...
+    other_stuff(color.green);
+    ...
+    yet_more_stuff(color.blue);
+}
+```
+
+To remedy the ergonomics problem, you may instead opt for using a positional
+tuple to contain the values you want to pass. However, now you have likely
+created a regression for readers of your code since the fields of tuples are
+accessed with their positional index, which does not carry clear semantic intent:
+
+```rust
+fn caller() {
+    ...
+    let color = (255, 0, 0);
+    ...
+    do_stuff_with(color); // Unclear what each position means...
+    ...
+}
+
+fn do_stuff_with(color: (u8, u8, u8)) { // More ergonomic!
+    some_stuff(color.0); // But less readable... :(
+    ...
+    other_stuff(color.1);
+    ...
+    yet_more_stuff(color.2);
+}
```

Comment:
```rust
fn do_stuff_with((red, green, blue): (u8, u8, u8)) {
    some_stuff(red);
    other_stuff(green);
    yet_more_stuff(blue);
}
```

Reply:
Even better, with #2522:

```rust
fn do_stuff_with((red: u8, green: u8, blue: u8)) {
    some_stuff(red);
    other_stuff(green);
    yet_more_stuff(blue);
}
```

However, while if you write it in this way it is clear what each tuple component means, it is not clear at the call site...
```rust
let color = (255, 0, 0);
// I can guess that this is (red, green, blue) because that's the usual for "color"
// but it isn't clear in the general case.
```